K3S Release Documentation Update (#6127)
k3s-io/k3s: docs/release/k3s-release.md

k3s-io/k3s
```

# Before You Begin
You’ll be primarily using git and go. Git can be installed via the local package manager. Make sure Go is installed and configured correctly, utilizing a “gopath”. This can be set via an environment variable called GOPATH. eg. export GOPATH=”${HOME}/go”, typically.

## Clone and Setup Remotes
Clone from upstream then add k3s-io fork and your personal fork.
```sh
# initial clone
git clone --origin upstream \
  https://github.com/kubernetes/kubernetes.git \
  ${GOPATH}/src/github.com/kubernetes/kubernetes
 
cd ${GOPATH}/src/github.com/kubernetes/kubernetes
 
# export your github name if different from user
# add the k3s-io remote
git remote add k3s-io https://github.com/k3s-io/kubernetes.git
 
# fetch all remote branches and tags. If you receive a message saying that
# previous tags will be "clobbered", add --force to the command below.
git fetch --all --tags
```
# Rebasing and Generating Tags
Establish a local branch for rebasing:

```sh
export GLOBAL_GIT_CONFIG_PATH=$(git config --list --show-origin --show-scope --global | awk 'NR==1{ split($2,path,":"); print path[2] }')
export SSH_MOUNT_PATH=$(echo $SSH_AUTH_SOCK || echo "$HOME/.ssh/id_rsa")

> # Set up your new/old versions of Kubernetes

Reviewer: Since we are removing from the assumptions from above, update the section below with: `<version>` or `<previous_version>` or the like where necessary. Here, and anywhere else below that references a specific version.

Author: This is done.

Reviewer: This needs to be completed in the rest of the doc.

git rebase --onto $NEW_K8S $OLD_K8S $OLD_K3S_VER~1
 
# Kubernetes is very picky about go versions. We use alpine and docker to specify the goversion with which we build the project.
# This command is not backwards compatible and requires versions of yq greater than 4.0, as the query syntax has changed throughout the history of the project.
export GOVERSION=$(yq -e '.dependencies[] | select(.name == "golang: upstream version").version' build/dependencies.yaml)

export GOIMAGE="golang:${GOVERSION}-alpine3.15"

export BUILD_CONTAINER="FROM ${GOIMAGE}\n \
RUN apk add --no-cache \
bash \
git \
make \
tar \
gzip \
curl \
git \
coreutils \
rsync \
alpine-sdk" 

echo -e $BUILD_CONTAINER | docker build -t ${GOIMAGE}-dev -

# Rebasing pulls in the tags.sh script.
# Now create the tags by executing tag.sh with the given version variables.
docker run --rm -u $(id -u) \
--mount type=tmpfs,destination=${GOPATH}/pkg \
-v ${GOPATH}/src:/go/src \
-v ${GOPATH}/.cache:/go/.cache \
-v ${GLOBAL_GIT_CONFIG_PATH}:/go/.gitconfig \
-e HOME=/go \
-e GOCACHE=/go/.cache \
-w /go/src/github.com/kubernetes/kubernetes ${GOIMAGE}-dev ./tag.sh ${NEW_K3S_VER} 2>&1 | tee ~/tags-${NEW_K3S_VER}.log
```

> After tag.sh runs, you should see list of git push commands at the end of the output.

Reviewer: ```suggestion
After tag.sh runs, you should see a list of `git push` commands at the end of the output.
```

# K3S Kubernetes Patch Release Process
This document details the K3S kubernetes patch release process.

> For reference, this document makes the following assumptions:

Reviewer: Does the doc need to make these assumptions? I think we can write it in such a way without the assumptions and provide examples where necessary.

Author: It doesn't have to make these assumption. However I put them in to provide context to the release. The origin document from which this is based greatly lacked context of this nature. Any thoughts on how to restructure to make this more implicit?

Reviewer: Remove the assumptions for now and that should be fine.

Author: This is done.

```
Current K3S Version: v1.22.11
New K3S Version [New k8s patch]: v1.22.12

Affected repos are:
k3s-io/kubernetes
k3s-io/k3s
```

# Before You Begin
You’ll be primarily using git and go. Git can be installed via the local package manager. Make sure Go is installed and configured correctly, utilizing a “gopath”. This can be set via an environment variable called GOPATH. eg. export GOPATH=”${HOME}/go”, typically.

## Clone and Setup Remotes
Clone from upstream then add k3s-io fork and your personal fork.
```sh
# initial clone
git clone --origin upstream \
  https://github.com/kubernetes/kubernetes.git \
  ${GOPATH}/src/github.com/kubernetes/kubernetes
 

> # cd into the repo

Reviewer: ```suggestion
```

Redundant

cd ${GOPATH}/src/github.com/kubernetes/kubernetes
 

> # export your github name if different from user

Reviewer 2: ```suggestion
```

There is no need for that since we are not using our personal forks anymore

Author: I'm happy to take this out although 2 releases prior we performed a release without personal forks and had to perform this section over to accommodate opening the pr from the users fork. What should we do here with the upcoming release in mind?

Author: This is done.

# add the k3s-io remote
git remote add k3s-io https://github.com/k3s-io/kubernetes.git
 
# fetch all remote branches and tags. If you receive a message saying that
# previous tags will be "clobbered", add --force to the command below.
git fetch --all --tags
```
# Rebasing and Generating Tags
Establish a local branch for rebasing:

```sh

> export GHUSER=<your-github-username>

Reviewer 2: ```suggestion
```
This is not needed, since we are not pushing to personal forks and not changing k3s mod to use it

> export GLOBAL_GITCONFIG_PATH=$(git config --list --show-origin --show-scope --global | awk 'NR==1{ split($2,path,":"); print path[2] }')

Reviewer: ```suggestion
export GLOBAL_GIT_CONFIG_PATH=$(git config --list --show-origin --show-scope --global | awk 'NR==1{ split($2,path,":"); print path[2] }')
```

rm -rf _output
 
# Rebase k3s customizations from the previous release, minus the merge commit (~1),
# onto the new upstream tag. This will leave you on a detached head that will be
# tagged in the following step.
git rebase --onto $NEW_K8S $OLD_K8S $OLD_K3S_VER~1
 
# Kubernetes is very picky about go versions. We use alpine and docker to specify the goversion with which we build the project.
# This command is not backwards compatible and requires versions of yq greater than 4.0, as the query syntax has changed throughout the history of the project.
export GOVERSION=$(yq -e '.dependencies[] | select(.name == "golang: upstream version").version' build/dependencies.yaml)

export GOIMAGE="golang:${GOVERSION}-alpine3.15"

export BUILD_CONTAINER="FROM ${GOIMAGE}\n \
RUN apk add --no-cache \
bash \
git \
make \
tar \
gzip \
curl \
git \
coreutils \
rsync \
alpine-sdk" 

echo -e $BUILD_CONTAINER | docker build -t ${GOIMAGE}-dev -

# Rebasing pulls in the tags.sh script.
# Now create the tags by executing tag.sh with the given version variables.
docker run --rm -u $(id -u) \
--mount type=tmpfs,destination=${GOPATH}/pkg \
-v ${GOPATH}/src:/go/src \
-v ${GOPATH}/.cache:/go/.cache \

> -v ${GLOBAL_GITCONFIG_PATH}:/go/.gitconfig \

Reviewer: ```suggestion
-v ${GLOBAL_GIT_CONFIG_PATH}:/go/.gitconfig \
```

```
export REMOTE=k3s-io
./push.sh
```
## Updating k3s with the new tags
You now have a collection of tagged kubernetes modules in your worktree. By updating go.mod in k3s to point at these modules we will then be prepared to open a PR for review.
```sh
cd $GOPATH/src/github.com/rancher/k3s
git remote add upstream https://github.com/k3s-io/k3s.git
git fetch upstream
git checkout -B $NEW_K3S_VER upstream/$RELEASE_BRANCH
git clean -xfd
 
sed -Ei "\|github.com/k3s-io/kubernetes| s|${OLD_K3S_VER}|${NEW_K3S_VER}|" go.mod
sed -Ei "s/k8s.io\/kubernetes v\S+/k8s.io\/kubernetes $NEW_K8S/" go.mod
sed -Ei "s/$OLD_K8S_CLIENT/$NEW_K8S_CLIENT/g" go.mod 
 
# since drone perform the builds and tests for the updated tags we no longer need to run make locally.
# We now update the go.sum by running go mod tidy:
go mod tidy
git add go.mod go.sum

git commit --all --signoff -m "Update to $NEW_K8S"
git push --set-upstream origin $NEW_K3S_VER 
```

Create a commit with all the changes, and push this upstream.
Create a PR to merge your branch into the corresponding release branch, and wait for CI to run tests on the PR. Make sure to create the PR against the associated release branch for this update.

> Once CI passes and you receive two approvals, you may now squash-merge the PR and then tag an RC.

Reviewer: ```suggestion
Once CI passes and you receive two approvals, you may now squash-merge the PR and then tag an RC after the merge to master CI run completes.
```

# Create a Release Candidate 
Releases are kicked off and created by tagging a new tag.
To create a new release in Github UI perform the following:

> 1. Set title and tag as v1.22.5-rc1+k3s1, target as release-1.22

Reviewer: ```suggestion
1. Set title and tag according to the release version you're working on. E.g. v1.22.5-rc1+k3s1.
```

2. Leave description blank.
3. Check the pre-release field.
4. Publish

The resulting run can be viewed here: 
[k3s-io/k3s Drone Dashboard](https://drone-publish.k3s.io/k3s-io/k3s)

# Create GA Release Candidate

> Once QA has verified after 24-72 hours that the RC is good (or that any fixes have been added in follow up RC candidates), it is time for the general release.

Reviewer: ```suggestion
Once QA has verified that the RC is good (or that any fixes have been added in follow up RC candidates), it is time for the general release.
```

1. Create new release in GH UI

> 2. Set title: v1.22.5+k3s1, add description with release notes. Leave the tag section blank.

Reviewer: Update to be more like the comment above.

Author: This is done.

Reviewer: Was the change pushed?

3. Check the pre-release field.
4. Save as draft until RC testing is complete.

> Once QA signs off on RC:

Reviewer: This section seems out of place.

Author: This is done.

1. Set tag to be created - should match title.
2. Ensure as pre-release is still checked.
3. Publish.

> Once CI has completed and artifacts are created:

Reviewer: This is only after a 24 hour waiting period.

Author: this is done.

1. Edit release to remove pre-release, and save.

The resulting run can be viewed here: 
[k3s-io/k3s Drone Dashboard](https://drone-publish.k3s.io/k3s-io/k3s)

# Create Release Images
The k3s-upgrade repository bundles a k3s binary and script that allows a user to upgrade to a new k3s release.

Go to the [k3s-upgrade repository](https://github.com/k3s-io/k3s-upgrade) and manually create a new tag for the release. This will kick off a build of the image. This process is normally automated, however this can fail. If the automation does fail, do the following:

1. Draft a new release
2. Enter the tag (e.g. v1.22.5-rc1+k3s1).
3. Check k3s and k3s-upgrade images Exist

This process will take some time but upon completion, the images will be listed here.

> The k3s images will be published here.

Author: This is done

> The upgrade images will be published here.

Reviewer: Link missing?

Author: This is done.
